File: ios/MPEngKeyboardSuggestion/Sources/MPEngKeyboardSuggestion/Internal/MPLearningManager.swift
// ============================================================
// MPLearningManager.swift
// ============================================================
// User behavior learning: unigram boost, bigram context, recency
// NO SINGLETON - initialize through MPKeyboardSuggestion
// ============================================================

import Foundation

public struct MPLearningStats: Sendable {
    public let words: Int
    public let bigrams: Int
}

final class MPLearningManager {
    
    private struct LearningData: Codable {
        var unigramCounts: [String: Int] = [:]
        var bigramCounts: [String: [String: Int]] = [:]
        var lastUsed: [String: Date] = [:]
    }
    
    private var data = LearningData()
    private var lastContext = ""
    private var isDirty = false
    private var saveWorkItem: DispatchWorkItem?
    
    private let saveQueue = DispatchQueue(label: "com.mp.keyboard.learning", qos: .utility)
    private let storageURL: URL
    
    var recencyDecayDays = 30.0
    var maxEntries = 5000
    
    init(storageDirectory: URL? = nil) {
        let directory = storageDirectory ?? FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        storageURL = directory.appendingPathComponent("mp_keyboard_learning.json")
        load()
    }
    
    // MARK: - Public API
    
    func recordSelection(_ word: String, context: String? = nil) {
        let lowerWord = word.lowercased()
        let contextWord = (context ?? lastContext).lowercased()
            .split(separator: " ").last.map(String.init) ?? ""
        
        data.unigramCounts[lowerWord, default: 0] += 1
        
        if !contextWord.isEmpty {
            if data.bigramCounts[contextWord] == nil {
                data.bigramCounts[contextWord] = [:]
            }
            data.bigramCounts[contextWord]![lowerWord, default: 0] += 1
        }
        
        data.lastUsed[lowerWord] = Date()
        lastContext = lowerWord
        
        isDirty = true
        scheduleSave()
        
        if data.unigramCounts.count > maxEntries {
            prune()
        }
    }
    
    func getBoostScore(for word: String, context: String) -> Double {
        let lowerWord = word.lowercased()
        let contextWord = context.lowercased().split(separator: " ").last.map(String.init) ?? ""
        
        var score = 0.0
        
        if let count = data.unigramCounts[lowerWord] {
            score += Double(count) * 10
        }
        
        if !contextWord.isEmpty,
           let bigramMap = data.bigramCounts[contextWord],
           let count = bigramMap[lowerWord] {
            score += Double(count) * 50
        }
        
        if let lastUsedDate = data.lastUsed[lowerWord] {
            let daysSince = Date().timeIntervalSince(lastUsedDate) / 86400
            let decay = pow(0.5, daysSince / recencyDecayDays)
            score *= (0.5 + 0.5 * decay)
        }
        
        return score
    }
    
    func getBigramSuggestions(for contextWord: String, limit: Int = 5) -> [(word: String, count: Int)] {
        guard let bigramMap = data.bigramCounts[contextWord.lowercased()] else {
            return []
        }
        return bigramMap.sorted { $0.value > $1.value }
            .prefix(limit)
            .map { ($0.key, $0.value) }
    }
    
    func updateContext(_ context: String) {
        lastContext = context
    }
    
    func getStats() -> MPLearningStats {
        let bigramCount = data.bigramCounts.values.reduce(0) { $0 + $1.count }
        return MPLearningStats(words: data.unigramCounts.count, bigrams: bigramCount)
    }
    
    func clear() {
        data = LearningData()
        try? FileManager.default.removeItem(at: storageURL)
    }
    
    // MARK: - Persistence
    
    private func load() {
        guard let jsonData = try? Data(contentsOf: storageURL),
              let decoded = try? JSONDecoder().decode(LearningData.self, from: jsonData) else {
            return
        }
        data = decoded
    }
    
    private func save() {
        guard isDirty else { return }
        saveQueue.async { [weak self] in
            guard let self = self else { return }
            if let encoded = try? JSONEncoder().encode(self.data) {
                try? encoded.write(to: self.storageURL)
            }
            self.isDirty = false
        }
    }
    
    private func scheduleSave() {
        saveWorkItem?.cancel()
        let workItem = DispatchWorkItem { [weak self] in
            self?.save()
        }
        saveWorkItem = workItem
        DispatchQueue.main.asyncAfter(deadline: .now() + 5, execute: workItem)
    }
    
    private func prune() {
        let cutoffDate = Date().addingTimeInterval(-90 * 86400)
        
        let toRemove = data.lastUsed.filter { $0.value < cutoffDate && (data.unigramCounts[$0.key] ?? 0) < 3 }.keys
        for word in toRemove {
            data.unigramCounts.removeValue(forKey: word)
            data.lastUsed.removeValue(forKey: word)
        }
        
        for (context, words) in data.bigramCounts {
            data.bigramCounts[context] = words.filter { $0.value >= 2 }
            if data.bigramCounts[context]?.isEmpty == true {
                data.bigramCounts.removeValue(forKey: context)
            }
        }
    }
}

File: android/MPEngKeyboardSuggestion/src/main/java/com/mp/keyboard/suggestion/internal/MPLearningManager.kt
// ============================================================
// MPLearningManager.kt
// ============================================================
// User behavior learning: unigram boost, bigram context, recency
// NO SINGLETON - initialize through MPKeyboardSuggestion
// ============================================================

package com.mp.keyboard.suggestion.internal

import android.content.Context
import android.content.SharedPreferences
import kotlinx.coroutines.*
import org.json.JSONObject
import kotlin.math.pow

data class MPLearningStats(
    val words: Int,
    val bigrams: Int
)

class MPLearningManager(context: Context, prefsName: String = "mp_keyboard_learning") {
    
    companion object {
        private const val UNIGRAM_KEY = "unigram"
        private const val BIGRAM_KEY = "bigram"
        private const val LAST_USED_KEY = "last_used"
    }
    
    private val prefs: SharedPreferences = context.getSharedPreferences(prefsName, Context.MODE_PRIVATE)
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    
    private val unigramCounts = mutableMapOf<String, Int>()
    private val bigramCounts = mutableMapOf<String, MutableMap<String, Int>>()
    private val lastUsed = mutableMapOf<String, Long>()
    
    private var lastContext = ""
    private var isDirty = false
    private var saveJob: Job? = null
    
    var recencyDecayDays = 30.0
    var maxEntries = 5000
    
    init { load() }
    
    // Public API
    
    fun recordSelection(word: String, context: String? = null) {
        val lowerWord = word.lowercase()
        val contextWord = (context ?: lastContext).lowercase().split(" ").lastOrNull() ?: ""
        
        unigramCounts[lowerWord] = (unigramCounts[lowerWord] ?: 0) + 1
        
        if (contextWord.isNotEmpty()) {
            if (!bigramCounts.containsKey(contextWord)) bigramCounts[contextWord] = mutableMapOf()
            bigramCounts[contextWord]!![lowerWord] = (bigramCounts[contextWord]!![lowerWord] ?: 0) + 1
        }
        
        lastUsed[lowerWord] = System.currentTimeMillis()
        lastContext = lowerWord
        
        isDirty = true
        scheduleSave()
        
        if (unigramCounts.size > maxEntries) prune()
    }
    
    fun getBoostScore(word: String, context: String): Double {
        val lowerWord = word.lowercase()
        val contextWord = context.lowercase().split(" ").lastOrNull() ?: ""
        
        var score = 0.0
        
        unigramCounts[lowerWord]?.let { score += it * 10.0 }
        
        if (contextWord.isNotEmpty()) {
            bigramCounts[contextWord]?.get(lowerWord)?.let { score += it * 50.0 }
        }
        
        lastUsed[lowerWord]?.let { timestamp ->
            val daysSince = (System.currentTimeMillis() - timestamp) / 86400000.0
            val decay = 0.5.pow(daysSince / recencyDecayDays)
            score *= (0.5 + 0.5 * decay)
        }
        
        return score
    }
    
    fun getBigramSuggestions(contextWord: String, limit: Int = 5): List<Pair<String, Int>> {
        return bigramCounts[contextWord.lowercase()]
            ?.entries
            ?.sortedByDescending { it.value }
            ?.take(limit)
            ?.map { it.key to it.value }
            ?: emptyList()
    }
    
    fun updateContext(context: String) { lastContext = context }
    
    fun getStats(): MPLearningStats {
        val bigramCount = bigramCounts.values.sumOf { it.size }
        return MPLearningStats(unigramCounts.size, bigramCount)
    }
    
    fun clear() {
        unigramCounts.clear()
        bigramCounts.clear()
        lastUsed.clear()
        prefs.edit().clear().apply()
    }
    
    fun destroy() { scope.cancel() }
    
    // Private
    
    private fun load() {
        try {
            prefs.getString(UNIGRAM_KEY, null)?.let { json ->
                val obj = JSONObject(json)
                obj.keys().forEach { unigramCounts[it] = obj.getInt(it) }
            }
            
            prefs.getString(BIGRAM_KEY, null)?.let { json ->
                val obj = JSONObject(json)
                obj.keys().forEach { ctx ->
                    bigramCounts[ctx] = mutableMapOf()
                    val inner = obj.getJSONObject(ctx)
                    inner.keys().forEach { word -> bigramCounts[ctx]!![word] = inner.getInt(word) }
                }
            }
            
            prefs.getString(LAST_USED_KEY, null)?.let { json ->
                val obj = JSONObject(json)
                obj.keys().forEach { lastUsed[it] = obj.getLong(it) }
            }
        } catch (e: Exception) { }
    }
    
    private fun save() {
        if (!isDirty) return
        scope.launch {
            try {
                prefs.edit().apply {
                    putString(UNIGRAM_KEY, JSONObject(unigramCounts as Map<*, *>).toString())
                    
                    val bigramJson = JSONObject()
                    bigramCounts.forEach { (ctx, words) -> bigramJson.put(ctx, JSONObject(words as Map<*, *>)) }
                    putString(BIGRAM_KEY, bigramJson.toString())
                    
                    putString(LAST_USED_KEY, JSONObject(lastUsed as Map<*, *>).toString())
                    apply()
                }
                isDirty = false
            } catch (e: Exception) { }
        }
    }
    
    private fun scheduleSave() {
        saveJob?.cancel()
        saveJob = scope.launch {
            delay(5000)
            save()
        }
    }
    
    private fun prune() {
        val cutoff = System.currentTimeMillis() - 90 * 86400000L
        val toRemove = lastUsed.filter { it.value < cutoff && (unigramCounts[it.key] ?: 0) < 3 }.keys
        toRemove.forEach { unigramCounts.remove(it); lastUsed.remove(it) }
        
        bigramCounts.entries.toList().forEach { (ctx, words) ->
            bigramCounts[ctx] = words.filter { it.value >= 2 }.toMutableMap()
            if (bigramCounts[ctx]?.isEmpty() == true) bigramCounts.remove(ctx)
        }
    }
}
